File: C/beadsort/rosetta/rosetta.c
#include <stdio.h>
#include <stdlib.h>

void bead_sort(int *a, int len)
{
	int i, j, max, sum;
	unsigned char *beads;
#	define BEAD(i, j) beads[i * max + j]

	for (i = 1, max = a[0]; i < len; i++)
		if (a[i] > max) max = a[i];

	beads = calloc(1, max * len);

	/* mark the beads */
	for (i = 0; i < len; i++)
		for (j = 0; j < a[i]; j++)
			BEAD(i, j) = 1;

	for (j = 0; j < max; j++) {
		/* count how many beads are on each post */
		for (sum = i = 0; i < len; i++) {
			sum += BEAD(i, j);
			BEAD(i, j) = 0;
		}
		/* mark bottom sum beads */
		for (i = len - sum; i < len; i++) BEAD(i, j) = 1;
	}

	for (i = 0; i < len; i++) {
		for (j = 0; j < max && BEAD(i, j); j++);
		a[i] = j;
	}
	free(beads);
}

int main()
{
    int n = 1000;
    int x[n];
    srand(time(NULL));

    for (int i = 0; i < n; i++) {
        x[i] = rand() % 10000;  // Gera valores aleatórios entre 0 e 9999
    }

    bead_sort(x, n);

    return 0;
}

File: Java/beadsort/Rosetta/Rosetta.java
public class Rosetta 
{
    public static void main(String[] args)
    {
        Rosetta now = new Rosetta();
        int[] arr = new int[1000];
        for(int i = 0; i < arr.length; i++)
            arr[i] = (int)(Math.random() * 901) + 100; // Gera números aleatórios entre 100 e 1000
        
        int[] sort = now.beadSort(arr);
    }

    int[] beadSort(int[] arr)
    {
        int max = arr[0];
        for(int i = 1; i < arr.length; i++)
            if(arr[i] > max)
                max = arr[i];
        
        // Set up abacus
        char[][] grid = new char[arr.length][max];
        int[] levelcount = new int[max];
        for(int i = 0; i < max; i++)
        {
            levelcount[i] = 0;
            for(int j = 0; j < arr.length; j++)
                grid[j][i] = '_';
        }
        
        // Drop the beads
        for(int i = 0; i < arr.length; i++)
        {
            int num = arr[i];
            for(int j = 0; num > 0; j++)
            {
                grid[levelcount[j]++][j] = '*';
                num--;
            }
        }

        // Count the beads
        int[] sorted = new int[arr.length];
        for(int i = 0; i < arr.length; i++)
        {
            int putt = 0;
            for(int j = 0; j < max && grid[arr.length - 1 - i][j] == '*'; j++)
                putt++;
            sorted[i] = putt;
        }
        
        return sorted;
    }
}
